drivers/i3c/i3c_stm32.c

```diff
@@ -0,0 +1,2163 @@
+			LOG_INF("IBI done, payload received :%d,%d,%d\n", data->ibi_payload,
+				data->ibi_payload_size, data->ibi_target_addr);
+			if ((data->ibi_payload != 0) && (data->ibi_payload_size != 0)) {
+				struct i3c_device_desc *target;
+
+				target = i3c_dev_list_i3c_addr_find(dev, data->ibi_target_addr);
+
+				if (target != NULL) {
+					if (i3c_ibi_work_enqueue_target_irq(
+						    target, (uint8_t *)&data->ibi_payload,
+						    data->ibi_payload_size) != 0) {
+						LOG_ERR("Error enqueue IBI IRQ work");
+					}
+				} else {
+					LOG_ERR("IBI from unknown device addr 0x%x",
+						data->ibi_target_addr);
+				}
+			}
+		}
+	}
+
+	if (LL_I3C_IsActiveFlag_HJ(i3c)) {
+		int ret;
+
+		LL_I3C_ClearFlag_HJ(i3c);
+
+		ret = i3c_ibi_work_enqueue_hotjoin(dev);
+		if (ret != 0) {
+			LOG_ERR("IBI Failed to enqueue hotjoin work");
+		}
+	}
+
+	k_sem_give(&data->ibi_lock_sem);
+
+#endif
+
+	if (LL_I3C_IsActiveFlag_WKP(i3c)) {
+		LL_I3C_ClearFlag_WKP(i3c);
+	}
+}
+
+/* Handles the I3C error ISR */
+static void i3c_stm32_error_isr(void *arg)
+{
+	const struct device *dev = (const struct device *)arg;
+
+	const struct i3c_stm32_config *config = dev->config;
+	struct i3c_stm32_data *data = dev->data;
+	I3C_TypeDef *i3c = config->i3c;
+
+	i3c_stm32_log_err_type(dev);
+
+	LL_I3C_ClearFlag_ERR(i3c);
+
+	data->msg_state = STM32_I3C_MSG_ERR;
+
+	k_sem_give(&data->device_sync_sem);
+
+#ifdef CONFIG_PM_DEVICE_RUNTIME
+	(void)pm_device_runtime_put(dev);
+#endif
+	pm_policy_state_lock_put(PM_STATE_SUSPEND_TO_IDLE, PM_ALL_SUBSTATES);
+}
+
+#ifdef CONFIG_I3C_USE_IBI
+
+int i3c_stm32_ibi_enable(const struct device *dev, struct i3c_device_desc *target)
+{
+	int ret = 0;
+	uint8_t idx;
+	I3C_TypeDef *i3c;
+	struct i3c_ccc_events i3c_events;
+	struct i3c_stm32_data *data = dev->data;
+	const struct i3c_stm32_config *config = dev->config;
+
+	i3c = config->i3c;
+	if (!i3c_device_is_ibi_capable(target)) {
+		return -EINVAL;
+	}
+
+	if (data->ibi.num_addr >= ARRAY_SIZE(data->ibi.addr)) {
+		/* No more free entries in the IBI table */
+		LOG_ERR("%s: no more free space in the IBI table", __func__);
+		return -ENOMEM;
+	}
+
+	for (idx = 0; idx < ARRAY_SIZE(data->ibi.addr); idx++) {
+		if (data->ibi.addr[idx] == target->dynamic_addr) {
+			LOG_ERR("%s: selected target is already in the list", __func__);
+			return -EINVAL;
+		}
+	}
+
+	if (data->ibi.num_addr > 0) {
+		for (idx = 0; idx < ARRAY_SIZE(data->ibi.addr); idx++) {
+			if (data->ibi.addr[idx] == 0U) {
+				break;
+			}
+		}
+
+		if (idx >= ARRAY_SIZE(data->ibi.addr)) {
+			LOG_ERR("Cannot support more IBIs");
+			return -ENOTSUP;
+		}
+
+	} else {
+		idx = 0;
+	}
+
+	data->ibi.addr[idx] = target->dynamic_addr;
+	data->ibi.num_addr += 1U;
+
+	/* Tell target to enable IBI */
+	i3c_events.events = I3C_CCC_EVT_INTR;
+	ret = i3c_ccc_do_events_set(target, true, &i3c_events);
+	if (ret != 0) {
+		LOG_ERR("Error sending IBI ENEC for 0x%02x (%d)", target->dynamic_addr, ret);
+	}
+
+	if (i3c_device_is_ibi_capable(target)) {
```

Comment:
above you check if it is not ibi capable and return an error code. so this statement will always ever be `true`. do you mean to check if it has a payload with an ibi especially for that `IBIAddData` argument with `i3c_ibi_has_payload`